perl: EOL releases: fix zlib and y2020 bugs in < @5.30

bernhardkaindl opened the pull request:
This pull request was completely rewritten. Closed to start fresh with pull request 25801.

#### Bug fixes for end of life releases:

- perl@:5.24.1 needs zlib@:1.2.8 - shown by more than a dozen tests
      https://rt.cpan.org/Public/Bug/Display.html?id=120134
    
- perl@:5.26.2 needs gdbm@:1.14.1 - shown by the test suite:
      https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133295

haampie commented:
I think the reason we unconditionally enable `gdbm` and `berkeley-db` is because you can't explicitly disable them, meaning that it'll autodetect system installs and link to those if found. Can you verify that? Generally I'd be a fan to disable all optional dependencies by default, and I remember looking at doing that for Perl too, so that OpenSSL would be faster to install with Spack; since the database-deps are still there, I think this means I couldn't make it work...

With regard to older versions: we may also just deprecate them instead of patching them -- they are considered 'legacy' anwyays. Note that 5.32 includes security fixes https://metacpan.org/release/XSAWYERX/perl-5.32.0/view/pod/perldelta.pod#Security that are not backported to 5.30.x (if I understand correctly)

Thanks for trying to make tests pass, that's very helpful!

haampie commented on var/spack/repos/builtin/packages/perl/package.py:
```diff
@@ -284,9 +308,13 @@ def setup_build_environment(self, env):
-        env.set('ZLIB_INCLUDE', spec['zlib'].prefix.include)
-        env.set('ZLIB_LIB', spec['zlib'].libs.directories[0])
+        # The test suite of the End of life releases fails with our zlib
+        if self.version >= Version('5.30.0'):
```
Do you know why? It'd be better to have non-vendored dependencies if anything links to libperl (if it exists) and also requires zlib.

Do they require an old zlib? Do they patch zlib?

haampie replied:
If zlib follows semver they haven't had a breaking change since 2002...

haampie replied:
Also: can it be that the test itself is broken?

bernhardkaindl replied:
> Do you know why? It'd be better to have non-vendored dependencies if anything links to libperl (if it exists) and also requires zlib.
> Do they require an old zlib? Do they patch zlib?

I no idea on any of this, found a better solution using zlib@:1.2.8 for these, so this comment is obsolete now I think.

> If zlib follows semver they haven't had a breaking change since 2002...

Sorry for getting off-the rails on answering this comment, but that's not true as shown by these perl-zlib failures which are also produced in spack: https://rt.cpan.org/Public/Bug/Display.html?id=120134

AFAICS, semver is a much newer idea/proposal only used by relatively new software.

As you will see, zlib didn't "follow" semver, not in 2002, nor do many large project of that time that I know of (examples below). Also even if project do follow it, somtimes projects and changes are too large to be sure that no change indoduces a side effect and no bugfix breaks some expectation by some user.

Semver didn't exist back than or if it did't it was not something that was known. I worked at the headqarters of the 2nd-largest Linux distributor at that time. The company did a lot of in-house software development for that Linux distribution too and we would have known about if semver.org were popular.

Looking back, version number never were an indication of being free of regressions. Semver isn't pointless, but the idea that a movement or rule (also before semver) could dictate version compatibility for everything on the planet didn't appear to work.

The old dominant projects (and espcially huge complex projects like the Linux kernel, gcc, clang, perl, openssl, openssh, GNOME, KDE etc.) rather follow their own model which they established many many years ago, maybe see this:
https://en.wikipedia.org/wiki/Software_versioning#Change_significance
https://gist.github.com/jashkenas/cbd2b088e20279ae2c8e
The version practice of them seem to be rather based on the basic idea this guy describes as "Romantic versioning":
https://github.com/romversioning/romver

While the idea of having a strict rule like semver in place is in theory good, projects like large Linux have discovered, that they always had to adapt their numbering to what was most bearable and readable. For example: When the version numbers simply became too large because of many revisions then they changed the numbering scheme.

Again sorry for going off the rails, but I think it needs to be driven home that "semver everywhere" does not exist even if tried, unintendet bugs escape all the time.

> Also: can it be that the test itself is broken?

No, it's not one test. Sometimes some individual tests are broken, but not this many:
```py
# Looks like you failed 5 tests of 317.
     4412    #   Failed test at t/cz-03zlib-v1.t line 754.
     4413    #   Failed test at t/cz-03zlib-v1.t line 766.
     4414    #   Failed test at t/cz-03zlib-v1.t line 788.
     4415    #   Failed test at t/cz-03zlib-v1.t line 790.
# Looks like you failed 5 tests of 471
          got: '74154'
     ...
     4458    #     expected: '74154'
     4459    # Inflation Error: data error
     4460    #   Failed test 'got expected uncompressed data'
     4461    #   at t/cz-06gzsetp.t line 140.
     4462    # unc len = 247, input len = 74154

Looks like you failed 15 tests of 51.
```
- Especially a wrong test would not show data corruption, but fail because of some other issue.

I've tested and pushed the fix according to this link:

It's an issue which was fixed in 5.26: https://rt.cpan.org/Public/Bug/Display.html?id=120134
```py
-    depends_on('zlib+shared')
+    # :5.24.1 needs zlib@:1.2.8: https://rt.cpan.org/Public/Bug/Display.html?id=120134
+    depends_on('zlib+shared',        when='@5.26.2:')
+    depends_on('zlib+shared@:1.2.8', when='@5.20.3:5.24.1')
```

bernhardkaindl commented:
> I think the reason we unconditionally enable `gdbm` and `berkeley-db` is because you can't explicitly disable them, meaning that it'll autodetect system installs and link to those if found. Can you verify that? 

Did so: Actually, perl's INSTALL file documents it: https://github.com/Perl/perl5/blob/blead/INSTALL#L1074
```doc
By default, Configure will offer to build every extension which appears
to be supported.  For example, Configure will offer to build GDBM_File
only if it is able to find the gdbm library.

To disable certain extensions so that they are not built, use the
-Dnoextensions=... and -Donlyextensions=... options.  They both accept
a space-separated list of extensions, such as C<IPC/SysV>.
```
Tested using:
```py
    depends_on('gdbm')
...
         if '+gdbm' in spec:
            config_args.append('-Dextensions=GDBM_File')
        else:
            config_args.append('-Dnoextensions=GDBM_File')
```
A diff of the logfile shows that while it successfully tests compiling a test program, GDBM_File is not built.
But, NDBM_File and ODMB_File which also link using -lgdbm are compiled! ;-)

I found a report saying that gdbm-1.15 broke that test of perl-5.26.2, where gdbm-1.14.1 works: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133295

I tested it in spack, and this fixes that test of perl@5.26.2 indeed:
```py
-    depends_on('gdbm')
+    depends_on('gdbm@1.14.1')
```
I think it would be best to look at the individual tests and like this (for db, the error is: `undefined reference to `db_version'`),
and as an update, I pushed an update which replaces the gdbm changes with:

```py
-    depends_on('gdbm', when='+gdbm')
+    depends_on('gdbm',        when='@5.30.0:')
+    depends_on('gdbm@1.14.1', when='@:5.28.0')
```
I guess this is clear and clean for older EOL revisions, ok?

bernhardkaindl committed:
perl: Fix regressions in the end-of-life revisions

- Apply patches fixing the perl-y2020 bug (1970+50 years):
  https://github.com/Perl/perl5/issues/17410

- perl@:5.24.1 needs zlib@:1.2.8 - shown by more than a dozen tests
  ttps://rt.cpan.org/Public/Bug/Display.html?id=120134

- perl@:5.26.2 needs gdbm@:1.14.1 - shown by the test suite:
  https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133295

alalazo commented on var/spack/repos/builtin/packages/perl/package.py:
```diff
@@ -63,16 +64,29 @@ class Perl(Package):  # Perl doesn't use Autotools, it should subclass Package
     depends_on('berkeley-db')
     depends_on('bzip2+shared')
-    depends_on('zlib+shared')
+    # :5.24.1 needs zlib@:1.2.8: https://rt.cpan.org/Public/Bug/Display.html?id=120134
+    depends_on('zlib+shared',        when='@5.26.2:')
+    depends_on('zlib+shared@:1.2.8', when='@5.20.3:5.24.1')
```
The `+shared` constraint is orthogonal to the version constraint, so we can write those like this:
```suggestion
    depends_on('zlib+shared')
    # :5.24.1 needs zlib@:1.2.8: https://rt.cpan.org/Public/Bug/Display.html?id=120134
    depends_on('zlib@:1.2.8', when='@5.20.3:5.24.1')
```

alalazo commented on var/spack/repos/builtin/packages/perl/package.py:
```diff
@@ -63,16 +64,29 @@ class Perl(Package):  # Perl doesn't use Autotools, it should subclass Package
     extendable = True
 
-    depends_on('gdbm')
+    # :5.28 needs gdbm@:1:14.1: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133295
+    depends_on('gdbm',         when='@5.30.0:')
+    depends_on('gdbm@:1.14.1', when='@:5.28.0')
```
```suggestion
    depends_on('gdbm')
    # :5.28 needs gdbm@:1:14.1: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133295
    depends_on('gdbm@:1.14.1', when='@:5.28.0')
```

bernhardkaindl commented:
Closed to start fresh with pull request 25801.

alalazo reviewed (changes requested):
Thanks for volunteering to be a maintainer. I left a couple of minor comments. Let me know what you think.
